{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 36610 - Python and Unix for Bioinformaticians"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Project 6: Resistance to antibiotics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Details on how to read compressed files can be found [here](https://docs.python.org/3/library/gzip.html#gzip.decompress)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip # import gzip to decompress .gz files\n",
    "\n",
    "resistance_genes = open(\"data/resistance_genes_sub.fsa\", \"r\")\n",
    "reads1 = gzip.open(\"data/Unknown3_raw_reads_1.txt.gz\", \"rt\") # 'rt' -> read-in in text mode\n",
    "reads2 = gzip.open(\"data/Unknown3_raw_reads_2.txt.gz\", \"rt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Open subsets to increase speed\n",
    "ResFile = open(\"data/resistance_genes_sub.fsa\", \"r\")\n",
    "reads1 = open(\"data/Unknown3_raw_reads_1_sub.txt\", \"r\")\n",
    "outfile = open(\"data/results.txt\", \"w\")\n",
    "\n",
    "# define variables\n",
    "header  = \"\"   # will contain the header for each FASTA sequence\n",
    "sub_seq = \"\"   # i think this variable is not necessarily needed â€“ used to store part of the FASTA sequence\n",
    "seq     = []   # used to store entire FASTA sequence\n",
    "line_count = 3 # \n",
    "\n",
    "# k-mer length\n",
    "def kmer_generator(sequence, kmer_length):\n",
    "    kmer_list = list()\n",
    "    for i in range(0, len(sequence), 1):\n",
    "        if (i < len(sequence)-kmer_length+1):\n",
    "            kmer_list.append(sequence[i:i+kmer_length])\n",
    "    return(kmer_list)\n",
    "\n",
    "\n",
    "\n",
    "### resistance gene\n",
    "    ### - k-mers\n",
    "### -- read (use in; NO k-mer)\n",
    "    ### for each k-mer in read\n",
    "        ### count\n",
    "\n",
    "for line in ResFile:\n",
    "    line = line.rstrip()\n",
    "    if line.startswith(\">\"):\n",
    "        header = line\n",
    "        if sub_seq:\n",
    "            seq.append(sub_seq)\n",
    "            seq = \"\".join(seq)\n",
    "            #print(header + \"\\n\" + seq) # header + full sequence\n",
    "            # generate k-mers right here\n",
    "            res_kmer = kmer_generator(seq, 19) # k-mers of the resistence gene\n",
    "            print(header, file=outfile)\n",
    "            #print(res_kmer)\n",
    "            res_kmer_count = get_counter_list(res_kmer)\n",
    "            #print(res_kmer_count)\n",
    "            \n",
    "            for line in reads1:\n",
    "                line = line.rstrip()\n",
    "                if (line_count % 4 == 0): # prints every 4th line\n",
    "                    sequence = line\n",
    "                    # if res k-mer in read\n",
    "                    \n",
    "                    #read_kmer = kmer_generator(sequence, 19)\n",
    "                    #incr_counter_list(res_kmer_count, read_kmer)\n",
    "                line_count += 1\n",
    "            \n",
    "            for pos in range(len(res_kmer_count)):\n",
    "                print(res_kmer_count[pos][0], res_kmer_count[pos][1], file=outfile)\n",
    "            seq = []\n",
    "    else:\n",
    "        sub_seq = line\n",
    "        seq.append(sub_seq)\n",
    "# the following is to process the last FASTA entry\n",
    "seq.append(sub_seq)\n",
    "seq = \"\".join(seq)\n",
    "#print(header + \"\\n\" + seq)\n",
    "# generate k-mers right here\n",
    "res_kmer = kmer_generator(seq, 19) # k-mers of the resistence gene\n",
    "#print(header)\n",
    "#print(res_kmer)\n",
    "\n",
    "ResFile.close()\n",
    "reads1.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "RES_TEST  = [\"ABC\", \"DEF\", \"GHI\", \"ABC\"]\n",
    "READ_LIST_2 = [\"DEF\", \"GHI\"]\n",
    "\n",
    "def get_counter_list(RES_LIST):\n",
    "    counterlist = list()\n",
    "    for entry in RES_LIST:\n",
    "        counterlist.append([entry, 0])\n",
    "    return(counterlist)\n",
    "\n",
    "# counterlist = get_counter_list(RES_TEST) # print list with k-mer and its coverage\n",
    "\n",
    "def incr_counter_list(RES_LIST, READ_LIST):\n",
    "    for pos in range(len(RES_LIST)):\n",
    "        for read_kmer in READ_LIST:\n",
    "            if read_kmer == RES_LIST[pos][0]: # in\n",
    "                #print(\"DOM\")\n",
    "                RES_LIST[pos][1] += 1\n",
    "        #print(RES_LIST[pos][0], RES_LIST[pos][1])\n",
    "    #return(RES_LIST) # do not return to modify original list\n",
    "    \n",
    "incr_counter_list(get_counter_list(RES_TEST), READ_LIST_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "reads1 = open(\"data/Unknown3_raw_reads_1_sub.txt\")\n",
    "\n",
    "#sequence = \"\"\n",
    "#line_count = 3\n",
    "\n",
    "#def get_read1(x):\n",
    "for line in reads1:\n",
    "    line = line.rstrip()\n",
    "    if (line_count % 4 == 0): # prints every 4th line\n",
    "        sequence = line\n",
    "        #print(sequence)\n",
    "    line_count += 1\n",
    "    \n",
    "# same s#*t\n",
    "def get_read1(line):\n",
    "    line_count = 3\n",
    "    line = line.rstrip()\n",
    "    if (line_count % 4 == 0): # prints every 4th line\n",
    "        sequence = line\n",
    "        return(sequence)\n",
    "    line_count += 1\n",
    "\n",
    "reads1.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# k_mer of resistance genes\n",
    "def k_mer(seq, k_mer_length):\n",
    "    k_mer_list = list()\n",
    "    for i in range(0, len(seq), 1):\n",
    "        if (i < len(seq) - k_mer_length + 1):\n",
    "            k_mer_list.append(sequence[i:i+k_mer_length])\n",
    "    return(k_mer_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RES  = \"ACGTACGTACGT\" # Resistence gene\n",
    "SEQ1 = \"TTACGT\"       # Contains last 3 resistance bases (so only part)\n",
    "SEQ2 = \"TACGTACGT\"    # Entire sequence part of resistence gene\n",
    "SEQ3 = \"TACGTTA\"      # Contains first 4 resistence bases (first part)\n",
    "# Initialize list that keeps track of the depth\n",
    "seq_list_count = list()\n",
    "\n",
    "for i in range(len(RES)):\n",
    "    seq_list_count.append(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "Found match in first iteration. Investigating this SEQ sequence further.\n",
      "0 TAC 3 TAC\n",
      "Match\n",
      "1 ACG 4 ACG\n",
      "Match\n",
      "2 CGT 5 CGT\n",
      "Match\n",
      "3 GTT 6 GTA\n",
      "Mismatch\n",
      "Start in RES: 3 End in RES: 8\n",
      "A 0\n",
      "C 0\n",
      "G 0\n",
      "T 1\n",
      "A 1\n",
      "C 1\n",
      "G 1\n",
      "T 1\n",
      "A 0\n",
      "C 0\n",
      "G 0\n",
      "T 0\n"
     ]
    }
   ],
   "source": [
    "# k-mer length\n",
    "def kmer_generator(sequence, kmer_length):\n",
    "    kmer_list = list()\n",
    "    for i in range(0, len(sequence), 1):\n",
    "        if (i < len(sequence)-kmer_length+1):\n",
    "            kmer_list.append(sequence[i:i+kmer_length])\n",
    "    return(kmer_list)\n",
    "\n",
    "kmer_length = 3\n",
    "\n",
    "SEQ1_kmer = kmer_generator(SEQ1, kmer_length)\n",
    "SEQ2_kmer = kmer_generator(SEQ2, kmer_length)\n",
    "SEQ3_kmer = kmer_generator(SEQ3, kmer_length)\n",
    "RES_kmer  = kmer_generator(RES, kmer_length)\n",
    "\n",
    "# check first SEQ k-mer (idea: COMPLETE SEQ-RES MATCH)\n",
    "# for this method only the first k_mer of SEQ needs to be generated => shorter runtime\n",
    "def first_kmer_check(SEQ_kmers, RES_kmers):\n",
    "    for i in range(0, len(RES_kmers)-1, 1):\n",
    "        if (SEQ_to_check[0] == RES_to_check[i]):\n",
    "            print(\"Found match in first iteration. Investigating this SEQ sequence further.\")\n",
    "            return(True, i)\n",
    "    print(\"No initial match. Discarding this SEQ.\")\n",
    "    return(False, 0)\n",
    "\n",
    "SEQ_to_check = SEQ3_kmer\n",
    "RES_to_check = RES_kmer\n",
    "\n",
    "first_hit, start_pos = first_kmer_check(SEQ_to_check, RES_to_check)\n",
    "end_pos = 0\n",
    "\n",
    "i = 0\n",
    "j = start_pos\n",
    "\n",
    "length_SEQ = len(SEQ_to_check) - 1\n",
    "length_RES = len(RES_to_check) - 1\n",
    "\n",
    "is_mismatch = False\n",
    "\n",
    "if (first_hit):\n",
    "    while (i <= length_SEQ):\n",
    "        while (j <= length_RES):\n",
    "            print(i, SEQ_to_check[i], j, RES_to_check[j])\n",
    "            if (SEQ_to_check[i] == RES_to_check[j]):\n",
    "                print(\"Match\")\n",
    "                end_pos = j\n",
    "                j += 1\n",
    "                break\n",
    "            elif (SEQ_to_check[i] != RES_to_check[j]):\n",
    "                print(\"Mismatch\")\n",
    "                is_mismatch = True\n",
    "                break\n",
    "            j += 1\n",
    "        if (is_mismatch):\n",
    "            j = 0\n",
    "            break\n",
    "        i += 1\n",
    "\n",
    "print(\"Start in RES:\", start_pos, \"End in RES:\", start_pos+end_pos)\n",
    "\n",
    "for i in range(start_pos, start_pos+end_pos, 1):\n",
    "    seq_list_count[i] += 1\n",
    "\n",
    "for i in range(len(RES)):\n",
    "    print(RES[i], seq_list_count[i])\n",
    "\n",
    "# RES:  ACGTACGTACGT\n",
    "# SEQ2: ---TACGTTA--"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@ILLUMINA-3BDE4F_0027:2:1:9216:1663#GCCAAT/1\n",
      "CAGCCCGCCGATGGCTCCCACAAGNTGTATCTCTGTACAGGTGTTATCGGGAGAATACTTGATTGCATTGGAAAGCAGGTTACTGAAAGCACGTCGGAGCA\n",
      "+ILLUMINA-3BDE4F_0027:2:1:9216:1663#GCCAAT/1\n",
      "ZSGDVEMEQHNGGEGGSGJESSWVBO_]_]\\DYV\\\\`SZ_aXaa`ccccccacc[Wcccaaccccacc[cca_cacc_cccaWccc[cccacc[ccb[ca_\n",
      "@ILLUMINA-3BDE4F_0027:2:1:10636:1724#GCCAAT/1\n",
      "TGCACAGGTAGCCCCTACGCCGCGNATGAACGACCGGAAACGCCGTCACATGATGGCGAAACCAGCCGACAAACTCCGCTTCCAGCCGCTTCAACACCGCC\n",
      "+ILLUMINA-3BDE4F_0027:2:1:10636:1724#GCCAAT/1\n",
      "a^DG_DPGGDQDFFFFQFKD\\\\Y\\Bab]ac[NY[acaXa]cc_ccYccccccbccYcccaccac^acbWccccc_cHabb_aaa[[_Y__a^^Iaab[RY_\n",
      "@ILLUMINA-3BDE4F_0027:2:1:8882:2205#GCCAAT/1\n",
      "TCCGAATTAAGCGCATATTCTGGTNCACGACTTTGAAGTGTCGCCCTTTTTAACCTCGCGAAAGCGTTGACTGTTATCCGCCTGAGAAGCCTGGAAGATGA\n",
      "+ILLUMINA-3BDE4F_0027:2:1:8882:2205#GCCAAT/1\n",
      "__FH^FSFMK[ELGHHQDVH_]Z[BddddddGaadccY^acfffffffffffffffdfffdfcfffdacef_ffcccfcef`WWW^^W^addeccYccbb^\n",
      "@ILLUMINA-3BDE4F_0027:2:1:6687:2439#GCCAAT/1\n",
      "AGATTCGTCGGATATTTGACGTTCNCCAGGTGCACGTAGCCCATCAGGTGGTCGACGTACTCCTTCAGCACCGATATCTGGCGCAGGGTCCCGCGACGCGC\n",
      "+ILLUMINA-3BDE4F_0027:2:1:6687:2439#GCCAAT/1\n"
     ]
    }
   ],
   "source": [
    "# print first lines of compressed FASTQ file\n",
    "count = 0\n",
    "for line in reads1:\n",
    "    if (count < 15):\n",
    "        print(line.rstrip())\n",
    "        count = count + 1\n",
    "    else:\n",
    "        break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
