{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 36610 - Python and Unix for Bioinformaticians"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Project 6: Resistance to antibiotics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## May 4, 2019\n",
    "### Peter's discussion\n",
    "- res genes can look alike/similar\n",
    "- he uses .find()\n",
    "- check only part of the read, if suffienctly enough kmers match, then go through each kmer of that read\n",
    "- do NOT loop over the res gene dict (-> slow)\n",
    "\n",
    "#### Set of RES kmer's to quickly check if sequencing read needs further investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pseudocode\n",
    "#\n",
    "# for line in resistance_genes_file:\n",
    "#    get header and sequence and store in a list\n",
    "#    generate a set of kmers and a dictionary of kmer and a value of 0\n",
    "#\n",
    "# for line in sequencing_read_1 and 2:\n",
    "#     isolate read sequence\n",
    "#     check if at least 2 out 3 read kmers match the resistance gene set\n",
    "#     if so:\n",
    "#         for each kmer in the read sequence:\n",
    "#             if kmer in resistance gene dictionary:\n",
    "#                 increase its value by 1\n",
    "#     else:\n",
    "#         check if the reverse complement has at least 2 out of 3 matches\n",
    "#         if that is the case:\n",
    "#             generate kmers for the reverse complement, and increase matching ones by 1\n",
    "#     otherwise discard this read (and do not generate kmers for it step by step)\n",
    "#\n",
    "# for resistance_gene in list_of_resistance_gene_header_and_sequence:\n",
    "#     get the header\n",
    "#     get the sequence\n",
    "#     get the length of the sequence\n",
    "#     for each kmer in that sequence:\n",
    "#         add the depth of each kmer to a temporary list\n",
    "#         check if that addition would result in a coverage <95%\n",
    "#         if it passes:\n",
    "#             continue collecting kmer depths for that sequence\n",
    "#         if not:\n",
    "#             discard this sequence\n",
    "#     since >95% coverage, and check for >=10 depth:\n",
    "#         we have a winner: store gene name, coverage and minimum depth in a final list\n",
    "#\n",
    "# sort the list according to coverage and depth, and print it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: res_kmer.py <FASTA_resistance_genes> <FASTQ_read_file1.gz> <FASTQ_read_file2.gz> <set_length_of_kmer>\n",
      "\n",
      "Processing resistance genes...\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'kmer_length' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-e01c79385e6a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    118\u001b[0m             \u001b[0;31m# add them to a kmer dictionary with value 0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    119\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfull_seq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 120\u001b[0;31m                 \u001b[0mkmer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfull_seq\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mkmer_length\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    121\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfull_seq\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mkmer_length\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    122\u001b[0m                     \u001b[0mResKmerSet\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkmer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'kmer_length' is not defined"
     ]
    }
   ],
   "source": [
    "# Libraries\n",
    "import sys\n",
    "import gzip # Provides support for *.gzip files\n",
    "import time # Used for runtime measurement\n",
    "time_start = time.time() # start timer\n",
    "\n",
    "print(\"\\nProgram: Tool for identifying which resistance genes are in the sample.\")\n",
    "print(\"Version: 1.0\")\n",
    "print(\"\\nUsage: res_kmer.py <FASTA_resistance_genes> <FASTQ_read_file1.gz> <FASTQ_read_file2.gz> <length_of_kmer>\\n\")\n",
    "\n",
    "if len(sys.argv) == 1: # no command line argument\n",
    "    print(\"No command line arguments.\")\n",
    "    try:\n",
    "        ResFile = open(str(input(\"Path to FASTA resistance gene file: \")), \"r\")\n",
    "    except IOError as error:\n",
    "        print(\"Can't open file, reason:\", str(error))\n",
    "        sys.exit(1)\n",
    "    Reads1_path = str(input(\"Path to FASTQ (*.gz) read 1 file: \")) # Unknown3_raw_reads_1.txt.gz\n",
    "    Reads2_path = str(input(\"Path to FASTQ (*.gz) read 2 file: \")) # Unknown3_raw_reads_2.txt.gz\n",
    "    try:\n",
    "        Reads1File = gzip.open(Reads1_path, \"rt\")\n",
    "        Reads2File = gzip.open(Reads2_path, \"rt\")\n",
    "    except IOError as error:\n",
    "        print(\"Can't open file, reason:\", str(error))\n",
    "        sys.exit(1)\n",
    "    try:\n",
    "        kmer_length = int(input(\"kmer length as integer (the lower the slower, recommended is 19): \"))\n",
    "    except ValueError as error:\n",
    "        sys.stderr.write(\"Kmer is not an integer. Exiting.\")\n",
    "        sys.exit(1)\n",
    "elif len(sys.argv) == 5: # user provided all command line arguments\n",
    "    try:\n",
    "        ResFile = open(sys.argv[1], \"r\")\n",
    "        Reads1File = gzip.open(sys.argv[2], \"rt\")\n",
    "        Reads2File = gzip.open(sys.argv[3], \"rt\")\n",
    "    except IOError as error:\n",
    "        print(\"Can't open file, reason:\", str(error))\n",
    "        sys.exit(1)\n",
    "    try:\n",
    "        kmer_length = int(sys.argv[4])\n",
    "    except ValueError as error:\n",
    "        sys.stderr.write(\"Kmer is not an integer. Exiting.\")\n",
    "        sys.exit(1)\n",
    "else:\n",
    "    sys.stderr.write(\"Usage: res_kmer.py <FASTA_resistance_genes> <FASTQ_read_file1.gz> <FASTQ_read_file2.gz> <length_of_kmer>\\n\")\n",
    "    sys.exit(1)\n",
    "\n",
    "# Functions\n",
    "def reverse_complement(DNA_sequence):\n",
    "    \"\"\"Reverse complement a given DNA sequence\"\"\"\n",
    "    DNA_translation_table = str.maketrans(\"ACGT\", \"TGCA\")\n",
    "    rev_compliment = DNA_sequence.translate(DNA_translation_table)\n",
    "    return rev_compliment[::-1]\n",
    "\n",
    "def read_in_seq_reads(read, kmer_length):\n",
    "    \"\"\"Go through sequencing reads and check if at least 2 out of 3 kmer match \n",
    "    the resistance gene kmer set. If so, generate all kmers for that read and \n",
    "    increase the count for matching kmers. Also consider the reverse complement.\"\"\"\n",
    "    \n",
    "    # Variables\n",
    "    read_seq = \"\"\n",
    "    rev_read_seq = \"\"\n",
    "    read_kmer = \"\"\n",
    "    line_count = 3\n",
    "    \n",
    "    for read_line in read:\n",
    "        if (line_count % 4 == 0):\n",
    "            read_seq = read_line.rstrip() # arrived at read sequence\n",
    "            \n",
    "            # Generate 3 read kmers and check if at least 2 match to the\n",
    "            # resistance gene dictionary. If so, generate all kmers for\n",
    "            # that read, otherwise try the reverse complement. If the\n",
    "            # reverse complement does not have at least 2 matches, ignore\n",
    "            # this read.\n",
    "            read_kmer_set = set()\n",
    "            read_kmer_set.add(read_seq[1:1+kmer_length])\n",
    "            read_kmer_set.add(read_seq[41:41+kmer_length])\n",
    "            read_kmer_set.add(read_seq[81:81+kmer_length])\n",
    "            \n",
    "            if len(read_kmer_set.intersection(ResKmerSet)) >= 2:\n",
    "                for j in range(0, len(read_seq), 1):\n",
    "                    if (j < len(read_seq) - kmer_length + 1):\n",
    "                        read_kmer = read_seq[j:j+kmer_length]\n",
    "                        if read_kmer in ResKmerDict.keys():\n",
    "                            ResKmerDict[read_kmer] += 1\n",
    "            else:\n",
    "                rev_read_seq = reverse_complement(read_seq) # reverse complement\n",
    "\n",
    "                read_kmer_set = set()\n",
    "                read_kmer_set.add(rev_read_seq[1:1+kmer_length])\n",
    "                read_kmer_set.add(rev_read_seq[41:41+kmer_length])\n",
    "                read_kmer_set.add(rev_read_seq[81:81+kmer_length])\n",
    "\n",
    "                if len(read_kmer_set.intersection(ResKmerSet)) >= 2:\n",
    "                    for j in range(0, len(rev_read_seq), 1):\n",
    "                        if (j < len(rev_read_seq) - kmer_length + 1):\n",
    "                            read_kmer = rev_read_seq[j:j+kmer_length]\n",
    "                            if read_kmer in ResKmerDict.keys():\n",
    "                                ResKmerDict[read_kmer] += 1\n",
    "        line_count += 1\n",
    "\n",
    "print(\"Processing resistance genes...\")\n",
    "fasta = []            # Will contain resistance gene sequence temporarily\n",
    "store_fasta_seqs = [] # List of header and sequence of resistance file\n",
    "ResKmerSet  = set()   # Resistance genes kmer set\n",
    "ResKmerDict = dict()  # Unique resistance genes kmer dictionary {kmer1:0, kmer2:0, ...}\n",
    "\n",
    "for line in ResFile:\n",
    "    line = line.rstrip()\n",
    "    if line.startswith(\">\"):\n",
    "        if fasta:\n",
    "            full_seq = \"\".join(fasta) # Store entire resistance gene sequence\n",
    "            store_fasta_seqs.append(header)\n",
    "            store_fasta_seqs.append(full_seq)\n",
    "            \n",
    "            # Generate kmers of the resistance gene,\n",
    "            # add them to a kmer set, and\n",
    "            # add them to a kmer dictionary with value 0\n",
    "            for i in range(0, len(full_seq), 1):\n",
    "                kmer = full_seq[i:i+kmer_length]\n",
    "                if (i < len(full_seq) - kmer_length + 1):\n",
    "                    ResKmerSet.add(kmer)\n",
    "                if kmer not in ResKmerDict.keys():\n",
    "                    ResKmerDict[kmer] = 0\n",
    "            fasta = []\n",
    "        header = line\n",
    "    else:\n",
    "        sequence = line\n",
    "        fasta.append(sequence)\n",
    "\n",
    "# Process last resistance gene\n",
    "if fasta:\n",
    "    full_seq = \"\".join(fasta) # Store entire resistance gene sequence\n",
    "    store_fasta_seqs.append(header)\n",
    "    store_fasta_seqs.append(full_seq)\n",
    "            \n",
    "    # Generate kmers of the resistance gene,\n",
    "    # add them to a kmer set, and\n",
    "    # add them to a kmer dictionary with value 0\n",
    "    for i in range(0, len(full_seq), 1):\n",
    "        kmer = full_seq[i:i+kmer_length]\n",
    "        if (i < len(full_seq) - kmer_length + 1):\n",
    "            ResKmerSet.add(kmer)\n",
    "        if kmer not in ResKmerDict.keys():\n",
    "            ResKmerDict[kmer] = 0\n",
    "\n",
    "ResFile.close()\n",
    "\n",
    "print(\"Finished processing resistance gene file.\")\n",
    "\n",
    "print(\"Processing read file 1...\")\n",
    "read_in_seq_reads(Reads1File, kmer_length)\n",
    "Reads1File.close()\n",
    "\n",
    "print(\"Processing read file 2...\")\n",
    "read_in_seq_reads(Reads2File, kmer_length)\n",
    "Reads2File.close()\n",
    "\n",
    "print(\"Filtering resistance genes that do not match the requirements (>95% coverage, >=10 depth)...\")\n",
    "\n",
    "skip_sequence = True\n",
    "counter = 0\n",
    "sequence = \"\"\n",
    "final_result = list()\n",
    "\n",
    "for i in range(len(store_fasta_seqs)):\n",
    "    if (counter % 2 == 0):\n",
    "        # arrived at gene id\n",
    "        skip_sequence = True\n",
    "        header = store_fasta_seqs[i]\n",
    "    else:\n",
    "        # arrived at gene sequence\n",
    "        temp = list()\n",
    "        sequence = store_fasta_seqs[i]\n",
    "        len_of_sequence = len(sequence)\n",
    "        \n",
    "        # Generate kmers of the sequence.\n",
    "        # Add the depth to a kmer depth list, and check for coverage.\n",
    "        # As soon as the coverage drops below 95%, ignore the sequence.\n",
    "        # (= does not fulfill the requirements).\n",
    "        for j in range(0, len(sequence), 1):\n",
    "            if (j < len(sequence) - kmer_length + 1):\n",
    "                read_kmer = sequence[j:j+kmer_length]\n",
    "                depth_of_kmer = ResKmerDict[read_kmer]\n",
    "                temp.append(depth_of_kmer)\n",
    "                temp_coverage = 1 - temp.count(0) / len_of_sequence\n",
    "                if (temp_coverage < 0.95): # check coverage, stop once below 95%\n",
    "                    skip_sequence = False\n",
    "                    break\n",
    "        # Once we processed every kmer of the sequence and the coverage\n",
    "        # is still >95%, we check for minimum depth (>=10)\n",
    "        if skip_sequence and (min(temp) >= 10):\n",
    "            final_result.append([header, temp_coverage, min(temp)])\n",
    "    counter += 1\n",
    "\n",
    "# Sort results according to coverage and then minimum depth. Print results.\n",
    "final_result.sort(key=lambda x: (x[1], x[2]), reverse=True)\n",
    "print(\"Coverage [%]\\tMinimum depth\\tGene\")\n",
    "for i in range(len(final_result)):\n",
    "    print(\"{:.2f}\\t\\t{:d}\\t\\t{}\".format(final_result[i][1] * 100, final_result[i][2], final_result[i][0]))\n",
    "\n",
    "time_end = time.time()\n",
    "time_elapsed = time_end - time_start\n",
    "time_elapsed"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
