{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 36610 - Python and Unix for Bioinformaticians"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Project 6: Resistance to antibiotics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## May 4, 2019\n",
    "### Peter's discussion\n",
    "- res genes can look alike/similar\n",
    "- he uses .find()\n",
    "- check only part of the read, if suffienctly enough kmers match, then go through each kmer of that read\n",
    "- do NOT loop over the res gene dict (-> slow)\n",
    "\n",
    "#### Set of RES kmer's to quickly check if sequencing read needs further investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating RES set.\n",
      "Finished setting up RES set.\n",
      "Reading read file 1...\n",
      "Reading read file 2...\n",
      "Determining whether sequence fulfills requirements...\n",
      "Cov [%]\t\tMin depth\tGene\n",
      "100.00\t\t27\t\t>blaSHV-28_1_HM751101 Beta-lactam resistance:\n",
      "100.00\t\t24\t\t>strA_4_AF321551 Aminoglycoside resistance:Alternate name; aph(3'')-Ib\n",
      "100.00\t\t21\t\t>strB_1_M96392 Aminoglycoside resistance:Alternate name; aph(6)-Id\n",
      "100.00\t\t20\t\t>blaCTX-M-15_23_DQ302097 Beta-lactam resistance:Alternate name; UOE-1\n",
      "100.00\t\t19\t\t>catB4_1_EU935739 Phenicol resistance:\n",
      "100.00\t\t18\t\t>aac(3)-IIa_1_CP023555.1 Aminoglycoside resistance:\n",
      "100.00\t\t16\t\t>blaTEM-1B_1_JF910132 Beta-lactam resistance:Alternate name; RblaTEM-1\n",
      "100.00\t\t16\t\t>sul2_2_GQ421466 Sulphonamide resistance:\n",
      "100.00\t\t15\t\t>tet(A)_4_AJ517790 Tetracycline resistance:\n",
      "100.00\t\t14\t\t>aac(6')Ib-cr_1_DQ303918 Fluoroquinolone and aminoglycoside resistance:\n",
      "100.00\t\t14\t\t>blaOXA-1_1_J02967 Beta-lactam resistance:\n",
      "Finished.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "56.28929305076599"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gzip\n",
    "import time         # time measurement\n",
    "start = time.time() # time measurement\n",
    "\n",
    "def reverse_complement(DNA_sequence):\n",
    "    \"\"\"Reverse complement a given DNA sequence\"\"\"\n",
    "    DNA_translation_table = str.maketrans(\"ACGT\", \"TGCA\")\n",
    "    rev_compliment = DNA_sequence.translate(DNA_translation_table)\n",
    "    return rev_compliment[::-1]\n",
    "\n",
    "fasta = [] # will contain the full resistance gene sequence\n",
    "kmer_length = 19\n",
    "store_fasta_seqs = []\n",
    "\n",
    "print(\"Generating RES set.\")\n",
    "ResKmerSet  = set()\n",
    "ResKmerDict = dict()\n",
    "ResFile = open(\"data/resistance_genes.fsa\", \"r\")\n",
    "for line in ResFile:\n",
    "    line = line.rstrip()\n",
    "    if line.startswith(\">\"):\n",
    "        if fasta:\n",
    "            full_seq = \"\".join(fasta)\n",
    "            store_fasta_seqs.append(header)\n",
    "            store_fasta_seqs.append(full_seq)\n",
    "            for i in range(0, len(full_seq), 1):\n",
    "                kmer = full_seq[i:i+kmer_length]\n",
    "                if (i < len(full_seq) - kmer_length + 1):\n",
    "                    ResKmerSet.add(kmer)\n",
    "                if kmer not in ResKmerDict.keys():\n",
    "                    ResKmerDict[kmer] = 0\n",
    "            fasta = []\n",
    "        header = line\n",
    "    else:\n",
    "        sequence = line\n",
    "        fasta.append(sequence)\n",
    "\n",
    "# ToDo: process last resistance gene\n",
    "\n",
    "ResFile.close()\n",
    "\n",
    "print(\"Finished setting up RES set.\")\n",
    "print(\"Reading read file 1...\")\n",
    "\n",
    "line_count = 3\n",
    "reads1 = gzip.open(\"data/Unknown3_raw_reads_1.txt.gz\", \"rt\")\n",
    "for read_line in reads1:\n",
    "    if (line_count % 4 == 0):\n",
    "        # include reverse complement check and process\n",
    "        read_seq = read_line.rstrip()\n",
    "        read_kmer_set = set()\n",
    "        \n",
    "        read_kmer_set.add(read_seq[1:1+kmer_length])\n",
    "        read_kmer_set.add(read_seq[41:41+kmer_length])\n",
    "        read_kmer_set.add(read_seq[81:81+kmer_length])\n",
    "        \n",
    "        if len(read_kmer_set.intersection(ResKmerSet)) >= 2:\n",
    "            for j in range(0, len(read_seq), 1):\n",
    "                if (j < len(read_seq) - kmer_length + 1):\n",
    "                    read_kmer = read_seq[j:j+kmer_length]\n",
    "                    if read_kmer in ResKmerDict.keys():\n",
    "                        ResKmerDict[read_kmer] += 1\n",
    "                        \n",
    "        # reverse complement\n",
    "        rev_read_seq = reverse_complement(read_seq)\n",
    "        read_kmer_set = set()\n",
    "        \n",
    "        read_kmer_set.add(rev_read_seq[1:1+kmer_length])\n",
    "        read_kmer_set.add(rev_read_seq[41:41+kmer_length])\n",
    "        read_kmer_set.add(rev_read_seq[81:81+kmer_length])\n",
    "        \n",
    "        if len(read_kmer_set.intersection(ResKmerSet)) >= 2:\n",
    "            for j in range(0, len(rev_read_seq), 1):\n",
    "                if (j < len(rev_read_seq) - kmer_length + 1):\n",
    "                    read_kmer = rev_read_seq[j:j+kmer_length]\n",
    "                    if read_kmer in ResKmerDict.keys():\n",
    "                        ResKmerDict[read_kmer] += 1\n",
    "    line_count += 1\n",
    "    \n",
    "reads1.close()\n",
    "print(\"Reading read file 2...\")\n",
    "\n",
    "line_count = 3\n",
    "reads2 = gzip.open(\"data/Unknown3_raw_reads_2.txt.gz\", \"rt\")\n",
    "for read_line in reads2:\n",
    "    if (line_count % 4 == 0):\n",
    "        # include reverse complement check and process\n",
    "        read_seq = read_line.rstrip()\n",
    "        read_kmer_set = set()\n",
    "        \n",
    "        read_kmer_set.add(read_seq[1:1+kmer_length])\n",
    "        read_kmer_set.add(read_seq[41:41+kmer_length])\n",
    "        read_kmer_set.add(read_seq[81:81+kmer_length])\n",
    "        \n",
    "        if len(read_kmer_set.intersection(ResKmerSet)) >= 2:\n",
    "            for j in range(0, len(read_seq), 1):\n",
    "                if (j < len(read_seq) - kmer_length + 1):\n",
    "                    read_kmer = read_seq[j:j+kmer_length]\n",
    "                    if read_kmer in ResKmerDict.keys():\n",
    "                        ResKmerDict[read_kmer] += 1\n",
    "                        \n",
    "        # reverse complement\n",
    "        rev_read_seq = reverse_complement(read_seq)\n",
    "        read_kmer_set = set()\n",
    "        \n",
    "        read_kmer_set.add(rev_read_seq[1:1+kmer_length])\n",
    "        read_kmer_set.add(rev_read_seq[41:41+kmer_length])\n",
    "        read_kmer_set.add(rev_read_seq[81:81+kmer_length])\n",
    "        \n",
    "        if len(read_kmer_set.intersection(ResKmerSet)) >= 2:\n",
    "            for j in range(0, len(rev_read_seq), 1):\n",
    "                if (j < len(rev_read_seq) - kmer_length + 1):\n",
    "                    read_kmer = rev_read_seq[j:j+kmer_length]\n",
    "                    if read_kmer in ResKmerDict.keys():\n",
    "                        ResKmerDict[read_kmer] += 1\n",
    "    line_count += 1\n",
    "\n",
    "reads2.close()\n",
    "print(\"Determining whether sequence fulfills requirements...\")\n",
    "\n",
    "skip_sequence = True\n",
    "counter = 0\n",
    "sequence = \"\"\n",
    "final_result = list()\n",
    "for i in range(len(store_fasta_seqs)):\n",
    "    if (counter % 2 == 0):\n",
    "        skip_sequence = True\n",
    "        header = store_fasta_seqs[i]\n",
    "    else:\n",
    "        temp = list()\n",
    "        sequence = store_fasta_seqs[i]\n",
    "        len_of_sequence = len(sequence)\n",
    "        for j in range(0, len(sequence), 1):\n",
    "            if (j < len(sequence) - kmer_length + 1):\n",
    "                read_kmer = sequence[j:j+kmer_length]\n",
    "                depth_of_kmer = ResKmerDict[read_kmer]\n",
    "                temp.append(depth_of_kmer)\n",
    "                temp_coverage = 1 - temp.count(0) / len_of_sequence\n",
    "                if (temp_coverage < 0.95): # check coverage, stop once below 95%\n",
    "                    #print(\"BELOW 95\", temp_coverage)\n",
    "                    skip_sequence = False\n",
    "                    break\n",
    "        if skip_sequence and (min(temp) >= 10):\n",
    "            final_result.append([header, temp_coverage, min(temp)])\n",
    "    counter += 1\n",
    "    \n",
    "final_result.sort(key=lambda x: (x[1], x[2]), reverse=True)\n",
    "print(\"Cov [%]\\t\\tMin depth\\tGene\")\n",
    "for i in range(len(final_result)):\n",
    "    print(\"{:.2f}\\t\\t{:d}\\t\\t{}\".format(final_result[i][1] * 100, final_result[i][2], final_result[i][0]))\n",
    "\n",
    "print(\"Finished.\")\n",
    "end = time.time()\n",
    "elapsed = end - start\n",
    "elapsed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Below: tests of other approaches"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Details on how to read compressed files can be found [here](https://docs.python.org/3/library/gzip.html#gzip.decompress)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip # import gzip to decompress .gz files\n",
    "\n",
    "resistance_genes = open(\"data/resistance_genes.fsa\", \"r\")\n",
    "reads1 = gzip.open(\"data/Unknown3_raw_reads_1.txt.gz\", \"rt\") # 'rt' -> read-in in text mode\n",
    "reads2 = gzip.open(\"data/Unknown3_raw_reads_2.txt.gz\", \"rt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading reads...\n",
      "Set contains in total 8201883 read k-mers.\n",
      "Number of sequencing reads including reverse complements is: 11022032\n",
      "Done setting up sets.\n",
      "Determining coverage and depth...\n",
      "Cov\t\tAvg depth\tGene\n",
      "95.01\t\t19.36\t\t>blaTEM-97_1_AF397067 Beta-lactam resistance:Alternate name; blaTEM-34\n",
      "95.10\t\t15.22\t\t>sul2_6_FN995456 Sulphonamide resistance:\n",
      "95.19\t\t19.20\t\t>sul2_9_FJ197818 Sulphonamide resistance:\n",
      "95.31\t\t29.41\t\t>blaOXA-320_1_KF151169 Beta-lactam resistance:\n",
      "95.38\t\t19.07\t\t>aac(6')Ib-cr_2_EF636461 Fluoroquinolone and aminoglycoside resistance:\n",
      "95.43\t\t22.02\t\t>blaCTX-M-22_1_HM470254 Beta-lactam resistance:\n",
      "95.47\t\t26.98\t\t>blaTEM-104_1_AF516719 Beta-lactam resistance:\n",
      "95.70\t\t23.70\t\t>blaTEM-99_1_AF397066 Beta-lactam resistance:Alternate name; blaTEM-30\n",
      "95.89\t\t23.06\t\t>blaCTX-M-28_6_AJ549244 Beta-lactam resistance:\n",
      "95.93\t\t12.90\t\t>blaTEM-93_1_AJ318093 Beta-lactam resistance:\n",
      "96.05\t\t15.84\t\t>blaTEM-183_1_HQ529916 Beta-lactam resistance:\n",
      "96.05\t\t23.34\t\t>blaTEM-55_1_DQ286729 Beta-lactam resistance:\n",
      "96.12\t\t19.60\t\t>blaCTX-M-15_70_FJ815277 Beta-lactam resistance:Alternate name; UOE-1\n",
      "96.23\t\t18.79\t\t>blaCTX-M-89_1_FJ966096 Beta-lactam resistance:\n",
      "96.28\t\t15.95\t\t>blaTEM-54_1_AF104442 Beta-lactam resistance:\n",
      "96.35\t\t16.72\t\t>blaCTX-M-52_1_GU125667 Beta-lactam resistance:\n",
      "96.35\t\t20.06\t\t>blaCTX-M-54_1_DQ303459 Beta-lactam resistance:\n",
      "96.46\t\t20.18\t\t>blaCTX-M-72_1_AY847148 Beta-lactam resistance:\n",
      "96.52\t\t19.24\t\t>blaTEM-1A_4_HM749966 Beta-lactam resistance:Alternate name; RblaTEM-1\n",
      "96.58\t\t20.18\t\t>blaCTX-M-42_1_DQ061159 Beta-lactam resistance:\n",
      "96.63\t\t24.40\t\t>blaSHV-28_2_EU441172 Beta-lactam resistance:\n",
      "96.69\t\t22.37\t\t>blaCTX-M-3_2_EF437434 Beta-lactam resistance:Amino acid sequences of CTX-M-3 and CTX-M-133 are identical\n",
      "96.86\t\t16.31\t\t>blaTEM-116_7_AY425988 Beta-lactam resistance:\n",
      "96.86\t\t23.58\t\t>blaTEM-57_1_FJ405211 Beta-lactam resistance:\n",
      "96.95\t\t19.85\t\t>sul2_14_AJ514834 Sulphonamide resistance:\n",
      "96.98\t\t17.71\t\t>blaTEM-20_1_EU527189 Beta-lactam resistance:\n",
      "96.98\t\t20.18\t\t>blaTEM-132_1_AY491682 Beta-lactam resistance:\n",
      "96.98\t\t26.98\t\t>blaTEM-206_1_KC783461 Beta-lactam resistance:\n",
      "97.06\t\t15.72\t\t>sul2_1_AF542061 Sulphonamide resistance:\n",
      "97.15\t\t23.41\t\t>blaCTX-M-15_64_FJ815288 Beta-lactam resistance:Alternate name; UOE-1\n",
      "97.21\t\t19.00\t\t>blaTEM-171_1_GQ149347 Beta-lactam resistance:\n",
      "97.21\t\t23.93\t\t>blaTEM-1C_5_FJ560503 Beta-lactam resistance:Alternate name; RblaTEM-1\n",
      "97.33\t\t19.59\t\t>blaTEM-147_1_DQ279850 Beta-lactam resistance:\n",
      "97.33\t\t25.75\t\t>aac(6')Ib-cr_1_DQ303918 Fluoroquinolone and aminoglycoside resistance:\n",
      "97.44\t\t24.87\t\t>blaTEM-166_1_FJ197316 Beta-lactam resistance:\n",
      "97.47\t\t20.54\t\t>blaOXA-224_1_JN596991 Beta-lactam resistance:\n",
      "97.55\t\t17.33\t\t>sul2_11_AY232670 Sulphonamide resistance:\n",
      "97.56\t\t20.76\t\t>blaTEM-10_1_AF093512 Beta-lactam resistance:Alternate name; MGH-1, blaTEM-23, blaTEM-E3\n",
      "97.56\t\t24.52\t\t>blaTEM-207_1_KC818234 Beta-lactam resistance:\n",
      "97.68\t\t23.58\t\t>blaTEM-29_1_DQ269440 Beta-lactam resistance:\n",
      "97.68\t\t24.52\t\t>blaTEM-135_1_GQ896333 Beta-lactam resistance:\n",
      "97.83\t\t20.79\t\t>tet(A)_5_AJ419171 Tetracycline resistance:\n",
      "98.03\t\t24.99\t\t>blaTEM-126_1_AY628199 Beta-lactam resistance:\n",
      "98.10\t\t43.29\t\t>fosA_3_ACWO01000079 Fosfomycin resistance:\n",
      "98.14\t\t24.28\t\t>blaTEM-143_1_DQ075245 Beta-lactam resistance:\n",
      "98.37\t\t20.06\t\t>blaTEM-79_1_AF190692 Beta-lactam resistance:\n",
      "98.38\t\t21.86\t\t>strA_1_M96392 Aminoglycoside resistance:Alternate name; aph(3'')-Ib\n",
      "98.38\t\t21.86\t\t>strA_3_AF024602 Aminoglycoside resistance:Alternate name; aph(3'')-Ib\n",
      "98.40\t\t20.64\t\t>blaCTX-M-79_1_GU125666 Beta-lactam resistance:\n",
      "98.41\t\t17.33\t\t>sul2_18_AJ830714 Sulphonamide resistance:\n",
      "98.49\t\t18.89\t\t>blaTEM-52B_1_AF027199 Beta-lactam resistance:\n",
      "98.49\t\t20.41\t\t>blaTEM-77_1_AF190695 Beta-lactam resistance:Alternate name; IRT-21\n",
      "98.49\t\t21.70\t\t>blaTEM-15_1_AM849805 Beta-lactam resistance:\n",
      "98.49\t\t24.99\t\t>blaTEM-208_1_KC865667 Beta-lactam resistance:\n",
      "98.52\t\t20.41\t\t>blaCTX-M-114_1_GQ351346 Beta-lactam resistance:\n",
      "98.61\t\t25.69\t\t>blaSHV-1_15_FJ668804 Beta-lactam resistance:\n",
      "98.63\t\t21.23\t\t>strA_2_M28829 Aminoglycoside resistance:Alternate name; aph(3'')-Ib\n",
      "98.67\t\t20.96\t\t>tet(A)_6_AJ313332 Tetracycline resistance:\n",
      "98.69\t\t28.12\t\t>strB_1_M96392 Aminoglycoside resistance:Alternate name; aph(6)-Id\n",
      "98.72\t\t22.29\t\t>blaTEM-106_1_AY101578 Beta-lactam resistance:\n",
      "98.72\t\t24.28\t\t>blaTEM-70_1_AF188199 Beta-lactam resistance:\n",
      "98.72\t\t25.46\t\t>blaTEM-148_1_AM087454 Beta-lactam resistance:\n",
      "98.74\t\t20.29\t\t>blaCTX-M-101_1_HQ398214 Beta-lactam resistance:\n",
      "98.84\t\t20.88\t\t>blaTEM-28_1_ECU37195 Beta-lactam resistance:\n",
      "98.84\t\t22.41\t\t>blaSHV-55_1_DQ054528 Beta-lactam resistance:\n",
      "98.84\t\t23.81\t\t>blaTEM-33_1_GU371926 Beta-lactam resistance:Alternate name; IRT-5\n",
      "98.84\t\t23.81\t\t>blaTEM-34_1_KC292503 Beta-lactam resistance:Alternate name; IRT-6\n",
      "98.90\t\t16.21\t\t>sul2_17_U57647 Sulphonamide resistance:\n",
      "98.90\t\t16.34\t\t>sul2_15_FJ968160 Sulphonamide resistance:\n",
      "98.90\t\t17.95\t\t>sul2_13_AJ289135 Sulphonamide resistance:\n",
      "98.95\t\t24.16\t\t>blaTEM-95_1_AJ308558 Beta-lactam resistance:\n",
      "98.95\t\t24.40\t\t>blaTEM-105_1_AF516720 Beta-lactam resistance:\n",
      "98.95\t\t24.63\t\t>blaTEM-186_1_JN227084 Beta-lactam resistance:\n",
      "98.97\t\t20.29\t\t>blaCTX-M-33_1_AY238472 Beta-lactam resistance:\n",
      "98.97\t\t22.60\t\t>blaCTX-M-71_1_FJ815436 Beta-lactam resistance:\n",
      "98.97\t\t22.60\t\t>blaCTX-M-103_1_HG423149 Beta-lactam resistance:\n",
      "99.00\t\t26.26\t\t>tet(A)_4_AJ517790 Tetracycline resistance:\n",
      "99.07\t\t24.16\t\t>blaTEM-176_1_GU550123 Beta-lactam resistance:\n",
      "99.07\t\t24.75\t\t>blaTEM-30_1_AJ437107 Beta-lactam resistance:Alternate name; IRT-2, TRI-2, E-guer\n",
      "99.07\t\t24.75\t\t>blaTEM-76_1_AF190694 Beta-lactam resistance:Alternate name; IRT-20\n",
      "99.11\t\t22.33\t\t>blaSHV-100_1_AM941846 Beta-lactam resistance:\n",
      "99.19\t\t22.64\t\t>blaSHV-106_1_AM922307 Beta-lactam resistance:\n",
      "99.28\t\t21.76\t\t>blaOXA-4_1_AY162283 Beta-lactam resistance:\n",
      "99.30\t\t24.63\t\t>blaTEM-127_1_AY368236 Beta-lactam resistance:\n",
      "99.39\t\t16.96\t\t>sul2_12_AF497970 Sulphonamide resistance:\n",
      "99.39\t\t20.05\t\t>sul2_2_GQ421466 Sulphonamide resistance:\n",
      "99.42\t\t28.15\t\t>blaTEM-1B_1_JF910132 Beta-lactam resistance:Alternate name; RblaTEM-1\n",
      "99.43\t\t20.41\t\t>blaCTX-M-82_1_DQ256091 Beta-lactam resistance:\n",
      "99.43\t\t21.10\t\t>blaCTX-M-11_1_AJ310929 Beta-lactam resistance:\n",
      "99.50\t\t24.75\t\t>strA_4_AF321551 Aminoglycoside resistance:Alternate name; aph(3'')-Ib\n",
      "99.54\t\t26.04\t\t>blaSHV-28_1_HM751101 Beta-lactam resistance:\n",
      "99.64\t\t24.10\t\t>catB4_1_EU935739 Phenicol resistance:\n",
      "99.66\t\t20.98\t\t>blaCTX-M-55_2_GQ456159 Beta-lactam resistance:Amino acid sequences of CTX-M-55 and CTX-M-57 are identical\n",
      "99.66\t\t22.14\t\t>blaCTX-M-117_1_JN227085 Beta-lactam resistance:\n",
      "99.88\t\t26.75\t\t>aac(3)-IIa_1_CP023555.1 Aminoglycoside resistance:\n",
      "99.89\t\t24.44\t\t>blaCTX-M-15_23_DQ302097 Beta-lactam resistance:Alternate name; UOE-1\n",
      "100.00\t\t30.75\t\t>blaOXA-1_1_J02967 Beta-lactam resistance:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2291.5162189006805"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import time         # time measurement\n",
    "start = time.time() # time measurement\n",
    "\n",
    "k_mer_length = 19\n",
    "\n",
    "read_kmer_set = set()        # set will contain k-mers of sequencing reads\n",
    "all_sequencing_reads = set() # stores sequencing reads in set\n",
    "\n",
    "def read_in_seq_reads(read, kmer_length):\n",
    "    \"\"\"Extract sequencing read sequences and generate k-mers from them. Store in separate sets.\"\"\"\n",
    "    raw_reads = open(\"data/Unknown3_raw_\" + read + \".txt\", \"r\")\n",
    "    \n",
    "    # Variables\n",
    "    read_seq = \"\"\n",
    "    read_seq_rev_comp = \"\"\n",
    "    line_count = 3\n",
    "    \n",
    "    for read_line in raw_reads:\n",
    "        read_line = read_line\n",
    "        if (line_count % 4 == 0):                                # get sequence line\n",
    "            read_seq = read_line.rstrip()\n",
    "            read_seq_rev_comp = reverse_complement(read_seq)     # consider other strand (reverse complement)\n",
    "        if (read_seq is not \"\" or read_seq_rev_comp is not \"\"):\n",
    "            # store read k-mer and its complement taken from middle portion of seq in a set.\n",
    "            read_kmer_set.add(read_seq[44:44+kmer_length])\n",
    "            read_kmer_set.add(read_seq_rev_comp[44:44+kmer_length])\n",
    "            # store sequencing reads in a separate set ->loop only once through sequencing read file\n",
    "            all_sequencing_reads.add(read_seq)\n",
    "            all_sequencing_reads.add(read_seq_rev_comp)\n",
    "        line_count += 1\n",
    "    raw_reads.close()\n",
    "\n",
    "def reverse_complement(DNA_sequence):\n",
    "    \"\"\"Reverse complement a given DNA sequence\"\"\"\n",
    "    DNA_translation_table = str.maketrans(\"ACGT\", \"TGCA\")\n",
    "    rev_compliment = DNA_sequence.translate(DNA_translation_table)\n",
    "    return rev_compliment[::-1]\n",
    "\n",
    "def k_mer_set_of_RES(sequence, kmer_length):\n",
    "    '''Return a set of k-mers of given sequence.\n",
    "    Used to search the single read k-mer in this set'''\n",
    "    k_mer_set = set()\n",
    "    for i in range(0, len(sequence), 1):\n",
    "        if (i < len(sequence)-kmer_length+1):\n",
    "            k_mer_set.add(sequence[i:i+kmer_length])\n",
    "    return(k_mer_set)\n",
    "\n",
    "def find_them_all(read_seq, res_seq):\n",
    "    \"\"\"Find all positions of a read in a resistance gene\"\"\"\n",
    "    i = res_seq.find(read_seq)\n",
    "    while (i != -1):\n",
    "        yield i\n",
    "        i = res_seq.find(read_seq, i + 1) # increase counter to prevent duplicates\n",
    "\n",
    "def determine_coverage_depth(full_seq, kmer_intersection):\n",
    "    # initialize a list with length of resistance gene and set values to 0\n",
    "    sequence_depth = [0 for i in range(len(full_seq))]\n",
    "\n",
    "    #print(\"Be aware that does not mean the entire read aligns to the sequence. Can still result in 0's.\")\n",
    "\n",
    "    # For all sequencing reads: if its k-mer is in the intersection (thus matches a resistance gene kmer),\n",
    "    # get all position to which the full read can potentially align in the resistance gene.\n",
    "    # We will obtain the starting position of the alignment. Based on that we can increase\n",
    "    # the coverage depth for the region that is covered by the entire read.\n",
    "    for read_seq in all_sequencing_reads:\n",
    "        if (read_seq[44:44+k_mer_length] in kmer_intersection):\n",
    "            hit_positions = find_them_all(read_seq, full_seq) # read may align to multiple positions\n",
    "            for hit in hit_positions:\n",
    "                for i in range(hit, hit+len(read_seq), 1):\n",
    "                    sequence_depth[i] += 1\n",
    "    return sequence_depth\n",
    "\n",
    "def depth_coverage_filter(depth_list, len_of_seq):\n",
    "    \"\"\"95% sequence coverage, at least depth of 10\"\"\"\n",
    "    count_of_zeros = 0\n",
    "    sum_of_depth = 0\n",
    "    \n",
    "    # reject sequence if single depth value is below 10\n",
    "    for i in depth_list:\n",
    "    #    if (i < 10):\n",
    "    #        return False\n",
    "        sum_of_depth += i\n",
    "    \n",
    "    # this would be a way to calculate the coverage, but it's not necessary because a depth<10 already rejects the sequence?!\n",
    "    # hence we would never count bases with coverage == 0\n",
    "        if (i == 0):\n",
    "            count_of_zeros += 1\n",
    "    \n",
    "    # coverage (>95%)\n",
    "    coverage = (len_of_seq - count_of_zeros) / len_of_seq\n",
    "    #if (coverage < 0.95):\n",
    "    #    return False\n",
    "    \n",
    "    # depth (>10)\n",
    "    avg_depth = sum_of_depth / len_of_seq\n",
    "    #if (avg_depth < 10):\n",
    "    #    return False\n",
    "    \n",
    "    return(coverage, avg_depth)\n",
    "\n",
    "print(\"Reading reads...\")\n",
    "\n",
    "read_in_seq_reads(\"reads_1\", k_mer_length)\n",
    "read_in_seq_reads(\"reads_2\", k_mer_length)\n",
    "\n",
    "print(\"Set contains in total {} read k-mers.\".format(len(read_kmer_set)))\n",
    "print(\"Number of sequencing reads including reverse complements is: {}\".format(len(all_sequencing_reads)))\n",
    "print(\"Done setting up sets.\")\n",
    "print(\"Determining coverage and depth...\")\n",
    "\n",
    "# Variables\n",
    "fasta = []            # will temporarily store the sequence for each resistance gene\n",
    "final_result = list() # stores final output\n",
    "\n",
    "ResFile = open(\"data/resistance_genes.fsa\", \"r\")\n",
    "for line in ResFile:\n",
    "    line = line.rstrip()\n",
    "    \n",
    "    # Arrived at resistance gene header. Collect its sequence.\n",
    "    # Once we get to the next resistance gene entry, process the previous one.\n",
    "    if line.startswith(\">\"):\n",
    "        \n",
    "        # Once we have the complete resistance gene sequence, generate k-mers of it. \n",
    "        # Find those k-mers that are identical to the read k-mers (using set intersection)\n",
    "        if fasta:\n",
    "            full_seq = \"\".join(fasta)\n",
    "            res_kmer_set = k_mer_set_of_RES(full_seq, k_mer_length)\n",
    "            kmer_intersection = res_kmer_set.intersection(read_kmer_set)\n",
    "\n",
    "            # Following 'if' can probably be more stringent i.e. allow only a minimum number of k-mers\n",
    "            # that are in common, but I suspect it would NOT drastically speed up the program.\n",
    "            if (len(kmer_intersection) > 10):\n",
    "                \n",
    "                ## Idea - but too slow: get those reads whose k-mer was in the intersection\n",
    "                #reads_of_interest = [ele for ele in all_sequencing_reads for x in kmer_intersection if x in ele]\n",
    "                \n",
    "                sequence_depth = determine_coverage_depth(full_seq, kmer_intersection)\n",
    "                                \n",
    "                # Once we processed all the reads, we can apply the filter.\n",
    "                # 1) 95% coverage of the resistance gene\n",
    "                # 2) Coverage depth of at least 10\n",
    "                coverage, avg_depth = depth_coverage_filter(sequence_depth, len(sequence_depth))\n",
    "                \n",
    "                # If it passes the filter, store it in a separate list for final evaluation\n",
    "                if (coverage > 0.95 and avg_depth > 10):\n",
    "                    final_result.append([header, coverage, avg_depth])\n",
    "            fasta = []\n",
    "        header = line\n",
    "        #print(header)\n",
    "    else:\n",
    "        sequence = line\n",
    "        fasta.append(sequence)\n",
    "\n",
    "# process last FASTA entry\n",
    "if fasta:\n",
    "    full_seq = \"\".join(fasta)\n",
    "    res_kmer_set = k_mer_set_of_RES(full_seq, k_mer_length)\n",
    "    kmer_intersection = res_kmer_set.intersection(read_kmer_set)\n",
    "    if (len(kmer_intersection) > 10):\n",
    "        sequence_depth = determine_coverage_depth(full_seq, kmer_intersection)\n",
    "        coverage, avg_depth = depth_coverage_filter(sequence_depth, len(sequence_depth))\n",
    "        if (coverage > 0.95 and avg_depth > 10):\n",
    "            final_result.append([header, coverage, avg_depth])\n",
    "\n",
    "ResFile.close()\n",
    "\n",
    "# Sort by\n",
    "# 1) Coverage, then\n",
    "# 2) Depth\n",
    "final_result.sort(key=lambda x: (x[1], x[2]))\n",
    "print(\"Cov\\t\\tAvg depth\\tGene\")\n",
    "for i in range(len(final_result)):\n",
    "    print(\"{:.2f}\\t\\t{:.2f}\\t\\t{}\".format(final_result[i][1]*100, final_result[i][2], final_result[i][0]))\n",
    "\n",
    "end = time.time()\n",
    "elapsed = end - start\n",
    "elapsed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## New approach"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Generate a **resistance gene dictionary of kmers** (RES_GENE_DICT), which looks like this:\n",
    "<pre><code class=\"python\">{gene_header1:{kmer1_1:0, kmer1_2:0, kmer1_3:0}, gene_header2:{kmer2_1:0, kmer2_2:0, kmer2_3:0}, ...}</code></pre>\n",
    "\n",
    "2. Loop over **read files**: generate read kmers and check if in RES_GENE_DICT. If so, increase corresponding value by 1. **BUT:** This results in too many comparisons/checks because we loop over the read files (which contains 3.5M reads each) and generate kmers of it which are then compared to the kmers of the 2000 resistance genes in the dict.\n",
    "\n",
    "3. Determine **coverage** and **depth**, and print genes that pass the filter (>95% coverage, >10 depth). Sorted output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished setting up RES dictionary.\n",
      "Processing reads...\n"
     ]
    }
   ],
   "source": [
    "import time         # time measurement\n",
    "start = time.time() # time measurement\n",
    "\n",
    "ResFile = open(\"data/resistance_genes.fsa\", \"r\")\n",
    "\n",
    "fasta = [] # will contain the full resistance gene sequence\n",
    "kmer_length = 19\n",
    "RES_kmer_dict = {} # dictionary {header:{kmer1:count1, kmer2:count2, ...}}\n",
    "\n",
    "# !!!\n",
    "# If RES kmers are the same, they are represented by a single kmer in the dictionary? Problem?\n",
    "# !!!\n",
    "\n",
    "# Variables for read file\n",
    "read_seq = \"\"\n",
    "read_seq_rev_comp = \"\"\n",
    "line_count = 3\n",
    "\n",
    "for line in ResFile:\n",
    "    line = line.rstrip()\n",
    "    \n",
    "    # Arrived at resistance gene header. Collect its sequence.\n",
    "    # Once we get to the next resistance gene entry, process the previous one.\n",
    "    if line.startswith(\">\"):\n",
    "        \n",
    "        # Once we have the complete resistance gene sequence, generate k-mers of it.\n",
    "        # feed it into a dictionary\n",
    "        if fasta:\n",
    "            full_seq = \"\".join(fasta)\n",
    "            \n",
    "            RES_kmer_dict[header] = {}\n",
    "            \n",
    "            for i in range(0, len(full_seq), 1):\n",
    "                if (i < len(full_seq) - kmer_length + 1):\n",
    "                    RES_kmer_dict[header][full_seq[i:i+kmer_length]] = 0\n",
    "            \n",
    "            fasta = []\n",
    "        header = line\n",
    "    else:\n",
    "        sequence = line\n",
    "        fasta.append(sequence)\n",
    "        \n",
    "# process last resistance gene\n",
    "full_seq = \"\".join(fasta)\n",
    "RES_kmer_dict[header] = {}\n",
    "            \n",
    "for i in range(0, len(full_seq), 1):\n",
    "    if (i < len(full_seq) - kmer_length + 1):\n",
    "        RES_kmer_dict[header][full_seq[i:i+kmer_length]] = 0\n",
    "\n",
    "ResFile.close()\n",
    "\n",
    "print(\"Finished setting up RES dictionary.\")\n",
    "print(\"Processing reads...\")\n",
    "\n",
    "# Go through the read files\n",
    "reads1 = open(\"data/Unknown3_raw_reads_1.txt\", \"r\")\n",
    "\n",
    "for read_line in reads1:\n",
    "    if (line_count % 4 == 0):\n",
    "        read_seq = read_line.rstrip()\n",
    "        \n",
    "        for j in range(0, len(read_seq), 1):\n",
    "            if (j < len(read_seq) - kmer_length + 1):\n",
    "                read_kmer = read_seq[j:j+kmer_length]\n",
    "                \n",
    "                # dictionary of read kmers\n",
    "                \n",
    "                # CPU intensive task:\n",
    "                # for EACH kmer, go through the RES dict\n",
    "                #\n",
    "                # Alternative:\n",
    "                # Store read kmer in a set (= does not contain duplicate kmers), and then loop through RES dictionary?? but then you have to loop over it again which takes time...\n",
    "                #\n",
    "                for kmer_combo in RES_kmer_dict.values():\n",
    "                    if read_kmer in kmer_combo.keys():\n",
    "                        kmer_combo[read_kmer] += 1\n",
    "    line_count += 1\n",
    "\n",
    "print(\"Done processing reads.\")\n",
    "print(RES_kmer_dict)\n",
    "print(\"Filtering results...\")\n",
    "\n",
    "final_result = list()\n",
    "\n",
    "for i, j in RES_kmer_dict.items():\n",
    "    min_depth = 0\n",
    "    number_of_res_kmers = len(j)\n",
    "    \n",
    "    kmer_above_zero = len([v for k, v in j.items() if v > 0])\n",
    "    coverage = int(kmer_above_zero * 100 / number_of_res_kmers)\n",
    "    #print(\"Coverage: {}%\".format(coverage))\n",
    "    \n",
    "    if coverage > 95:\n",
    "        #avg_depth = sum([v for k, v in j.items()]) / number_of_res_kmers\n",
    "        min_depth = min([v for k, v in j.items()])\n",
    "        #print(\"Average depth: {}\".format(avg_depth))\n",
    "        if min_depth > 10:\n",
    "            final_result.append([i, coverage, avg_depth])\n",
    "\n",
    "final_result.sort(key=lambda x: (x[1], x[2]))\n",
    "print(\"Cov\\t\\tAvg depth\\tGene\")\n",
    "for i in range(len(final_result)):\n",
    "    print(\"{:.2f}\\t\\t{:.2f}\\t\\t{}\".format(final_result[i][1], final_result[i][2], final_result[i][0]))\n",
    "\n",
    "reads1.close()\n",
    "\n",
    "end = time.time()\n",
    "elapsed = end - start\n",
    "elapsed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load reads in memory too"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing reads...\n",
      "Done processing reads.\n",
      "Finished setting up RES dictionary.\n",
      "Filtering results...\n",
      "Cov\t\tMin depth\tGene\n",
      "100.00\t\t16.00\t\t>sul2_2_GQ421466 Sulphonamide resistance:\n",
      "100.00\t\t19.00\t\t>aac(3)-IIa_1_CP023555.1 Aminoglycoside resistance:\n",
      "100.00\t\t21.00\t\t>blaCTX-M-15_23_DQ302097 Beta-lactam resistance:Alternate name; UOE-1\n",
      "100.00\t\t26.00\t\t>blaTEM-1B_1_JF910132 Beta-lactam resistance:Alternate name; RblaTEM-1\n",
      "100.00\t\t27.00\t\t>strB_1_M96392 Aminoglycoside resistance:Alternate name; aph(6)-Id\n",
      "100.00\t\t27.00\t\t>blaOXA-1_1_J02967 Beta-lactam resistance:\n",
      "100.00\t\t29.00\t\t>aac(6')Ib-cr_1_DQ303918 Fluoroquinolone and aminoglycoside resistance:\n",
      "100.00\t\t29.00\t\t>strA_4_AF321551 Aminoglycoside resistance:Alternate name; aph(3'')-Ib\n",
      "100.00\t\t29.00\t\t>catB4_1_EU935739 Phenicol resistance:\n",
      "100.00\t\t29.00\t\t>tet(A)_4_AJ517790 Tetracycline resistance:\n",
      "100.00\t\t37.00\t\t>blaSHV-28_1_HM751101 Beta-lactam resistance:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1321.9995760917664"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import time         # time measurement\n",
    "start = time.time() # time measurement\n",
    "\n",
    "def reverse_complement(DNA_sequence):\n",
    "    \"\"\"Reverse complement a given DNA sequence\"\"\"\n",
    "    DNA_translation_table = str.maketrans(\"ACGT\", \"TGCA\")\n",
    "    rev_compliment = DNA_sequence.translate(DNA_translation_table)\n",
    "    return rev_compliment[::-1]\n",
    "\n",
    "fasta = [] # will contain the full resistance gene sequence\n",
    "kmer_length = 19\n",
    "RES_dict_from_read_kmers = {} # dictionary {header:{kmer1:count1, kmer2:count2, ...}}\n",
    "\n",
    "# Variables for read file\n",
    "read_seq = \"\"\n",
    "read_seq_rev_comp = \"\"\n",
    "line_count = 3\n",
    "\n",
    "print(\"Processing reads...\")\n",
    "\n",
    "# Go through the read files\n",
    "read_kmer_dict = {}\n",
    "\n",
    "reads1 = open(\"data/Unknown3_raw_reads_1.txt\", \"r\")\n",
    "\n",
    "for read_line in reads1:\n",
    "    if (line_count % 4 == 0):\n",
    "        read_seq = read_line.rstrip()\n",
    "        \n",
    "        for j in range(0, len(read_seq), 1):\n",
    "            if (j < len(read_seq) - kmer_length + 1):\n",
    "                read_kmer = read_seq[j:j+kmer_length]\n",
    "                rev_read_kmer = reverse_complement(read_kmer)\n",
    "                \n",
    "                if read_kmer in read_kmer_dict.keys():\n",
    "                    read_kmer_dict[read_kmer] += 1\n",
    "                else:\n",
    "                    read_kmer_dict[read_kmer] = 1\n",
    "                \n",
    "                if rev_read_kmer in read_kmer_dict.keys():\n",
    "                    read_kmer_dict[rev_read_kmer] += 1\n",
    "                else:\n",
    "                    read_kmer_dict[rev_read_kmer] = 1\n",
    "    line_count += 1\n",
    "    \n",
    "line_count = 3\n",
    "reads2 = open(\"data/Unknown3_raw_reads_2.txt\", \"r\")\n",
    "for read_line in reads2:\n",
    "    if (line_count % 4 == 0):\n",
    "        read_seq = read_line.rstrip()        \n",
    "        for j in range(0, len(read_seq), 1):\n",
    "            if (j < len(read_seq) - kmer_length + 1):\n",
    "                read_kmer = read_seq[j:j+kmer_length]\n",
    "                rev_read_kmer = reverse_complement(read_kmer)\n",
    "                \n",
    "                if read_kmer in read_kmer_dict.keys():\n",
    "                    read_kmer_dict[read_kmer] += 1\n",
    "                else:\n",
    "                    read_kmer_dict[read_kmer] = 1\n",
    "                \n",
    "                if rev_read_kmer in read_kmer_dict.keys():\n",
    "                    read_kmer_dict[rev_read_kmer] += 1\n",
    "                else:\n",
    "                    read_kmer_dict[rev_read_kmer] = 1\n",
    "    line_count += 1\n",
    "\n",
    "reads1.close()\n",
    "reads2.close()\n",
    "print(\"Done processing reads.\")\n",
    "\n",
    "ResFile = open(\"data/resistance_genes.fsa\", \"r\")\n",
    "\n",
    "for line in ResFile:\n",
    "    line = line.rstrip()\n",
    "    \n",
    "    # Arrived at resistance gene header. Collect its sequence.\n",
    "    # Once we get to the next resistance gene entry, process the previous one.\n",
    "    if line.startswith(\">\"):\n",
    "        \n",
    "        # Once we have the complete resistance gene sequence, generate k-mers of it.\n",
    "        # feed it into a dictionary\n",
    "        if fasta:\n",
    "            full_seq = \"\".join(fasta)\n",
    "            \n",
    "            RES_dict_from_read_kmers[header] = {}\n",
    "            \n",
    "            for i in range(0, len(full_seq), 1):\n",
    "                if (i < len(full_seq) - kmer_length + 1):\n",
    "                    RES_kmer = full_seq[i:i+kmer_length]\n",
    "                    if RES_kmer in read_kmer_dict:\n",
    "                        RES_dict_from_read_kmers[header][RES_kmer] = read_kmer_dict[RES_kmer]\n",
    "                    else:\n",
    "                        RES_dict_from_read_kmers[header][RES_kmer] = 0\n",
    "            \n",
    "            fasta = []\n",
    "        header = line\n",
    "    else:\n",
    "        sequence = line\n",
    "        fasta.append(sequence)\n",
    "        \n",
    "# ToDo: Process last entry\n",
    "        \n",
    "ResFile.close()\n",
    "\n",
    "print(\"Finished setting up RES dictionary.\")\n",
    "\n",
    "#print(RES_dict_from_read_kmers)\n",
    "\n",
    "print(\"Filtering results...\")\n",
    "\n",
    "final_result = list()\n",
    "\n",
    "for i, j in RES_dict_from_read_kmers.items():\n",
    "    min_depth = 0\n",
    "    number_of_res_kmers = len(j)\n",
    "    \n",
    "    kmer_above_zero = len([v for k, v in j.items() if v > 0])\n",
    "    coverage = int(kmer_above_zero * 100 / number_of_res_kmers)\n",
    "    \n",
    "    if coverage > 95:\n",
    "        min_depth = min([v for k, v in j.items()])\n",
    "        if min_depth > 10:\n",
    "            final_result.append([i, coverage, min_depth])\n",
    "\n",
    "final_result.sort(key=lambda x: (x[1], x[2]))\n",
    "print(\"Cov\\t\\tMin depth\\tGene\")\n",
    "for i in range(len(final_result)):\n",
    "    print(\"{:.2f}\\t\\t{:.2f}\\t\\t{}\".format(final_result[i][1], final_result[i][2], final_result[i][0]))\n",
    "\n",
    "end = time.time()\n",
    "elapsed = end - start\n",
    "elapsed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test area"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_items([('head1', {'key1_1': 2, 'key1_2': 0, 'key1_3': 4}), ('head2', {'key2_1': 20, 'key2_2': 22})])\n",
      "Nope\n",
      "{'key1_1': 2, 'key1_2': 0, 'key1_3': 4}\n",
      "{'key2_1': 20, 'key2_2': 22}\n",
      "YES IT IS\n",
      "{'head1': {'key1_1': 3, 'key1_2': 0, 'key1_3': 4}, 'head2': {'key2_1': 20, 'key2_2': 22}}\n",
      "\n",
      "\n",
      "{'key1_1': 3, 'key1_2': 0, 'key1_3': 4}\n",
      "Average depth: 2.3333333333333335\n",
      "Coverage: 66%\n",
      "{'key2_1': 20, 'key2_2': 22}\n",
      "Average depth: 21.0\n",
      "Coverage: 100%\n",
      "Cov\t\tAvg depth\tGene\n",
      "100.00\t\t21.00\t\thead2\n"
     ]
    }
   ],
   "source": [
    "test_dict = {\"head1\":{\"key1_1\":2, \"key1_2\":0, \"key1_3\":4}, \"head2\":{\"key2_1\":20, \"key2_2\":22}}\n",
    "print(test_dict.items())\n",
    "# test_dict[\"head1\"][\"key1\"] += 1\n",
    "\n",
    "# test 1\n",
    "if \"key1\" in test_dict:\n",
    "    print(\"Yes\")\n",
    "else:\n",
    "    print(\"Nope\")\n",
    "\n",
    "# test 2\n",
    "for k in test_dict.values():\n",
    "    print(k)\n",
    "\n",
    "# test 3\n",
    "for j in test_dict.values():\n",
    "    if \"key1_1\" in j.keys():\n",
    "        j[\"key1_1\"] += 1\n",
    "\n",
    "print(test_dict)\n",
    "print(\"\\n\")\n",
    "\n",
    "final_result = list()\n",
    "\n",
    "for i,j in test_dict.items():\n",
    "    print(j)\n",
    "    number_of_res_kmers = len(j)\n",
    "    #print(number_of_res_kmers)\n",
    "    \n",
    "    avg_depth = sum([v for k, v in j.items()]) / number_of_res_kmers\n",
    "    print(\"Average depth: {}\".format(avg_depth))\n",
    "    \n",
    "\n",
    "    kmer_above_zero = len([v for k, v in j.items() if v > 0])\n",
    "    coverage = int(kmer_above_zero * 100 / number_of_res_kmers)\n",
    "    print(\"Coverage: {}%\".format(coverage))\n",
    "\n",
    "    if avg_depth > 10 or coverage > 95:\n",
    "        final_result.append([i, coverage, avg_depth])\n",
    "\n",
    "final_result.sort(key=lambda x: (x[1], x[2]))\n",
    "print(\"Cov\\t\\tAvg depth\\tGene\")\n",
    "for i in range(len(final_result)):\n",
    "    print(\"{:.2f}\\t\\t{:.2f}\\t\\t{}\".format(final_result[i][1], final_result[i][2], final_result[i][0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'El Salvador', 'Nicaragua'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importers = {'El Salvador' : 1234,\n",
    "             'Nicaragua' : 152,\n",
    "             'Spain' : 252\n",
    "            }\n",
    "\n",
    "exporters = {'Spain' : 252,\n",
    "             'Germany' : 251,\n",
    "             'Italy' : 1563\n",
    "             }\n",
    "importers.keys() - exporters.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
